services/kafka/hash.go

```diff
@@ -0,0 +1,116 @@
+
+type murmur2 struct{}
+
+// Partition here is based on the murmur2 balance function in github.com/segmentio/kafka-go to not break compatibility
+// with earlier versions.
+func (b murmur2) Partition(msg *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	if numPartitions == 0 {
+		return 0, ErrNonPositivePartitions
+	}
+
+	// NOTE: the murmur2 balancers in java and librdkafka treat a nil key as
+	//       non-existent while treating an empty slice as a defined value.
+	if msg.Key == nil {
+		return rand.New(rand.NewSource(time.Now().UTC().UnixNano())).Int31n(numPartitions), nil
+	}
+	key, err := msg.Key.Encode()
+	if err != nil {
+		return 0, err
+	}
+	return int32((murmur2Hash(key) & 0x7fffffff) % uint32(numPartitions)), nil
+}
+
+func (murmur2) RequiresConsistency() bool { return true }
+
+// murmur2 is taken from github.com/segmentio/kafka-go
+// We copy here because it was in a private method.
+// It is a go port of the Java library's murmur2 function.
+// https://github.com/apache/kafka/blob/1.0/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L353
+// Go port of the Java library's murmur2 function.
+// https://github.com/apache/kafka/blob/1.0/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L353
+func murmur2Hash(data []byte) uint32 {
+	length := len(data)
+	const (
+		seed uint32 = 0x9747b28c
+		// 'm' and 'r' are mixing constants generated offline.
+		// They're not really 'magic', they just happen to work well.
+		m = 0x5bd1e995
+		r = 24
+	)
+
+	// Initialize the hash to a random value
+	h := seed ^ uint32(length)
+	length4 := length / 4
+
+	for i := 0; i < length4; i++ {
+		i4 := i * 4
+		k := (uint32(data[i4+0]) & 0xff) + ((uint32(data[i4+1]) & 0xff) << 8) + ((uint32(data[i4+2]) & 0xff) << 16) + ((uint32(data[i4+3]) & 0xff) << 24)
+		k *= m
+		k ^= k >> r
+		k *= m
+		h *= m
+		h ^= k
+	}
+
+	// Handle the last few bytes of the input array
+	extra := length % 4
+	if extra >= 3 {
+		h ^= (uint32(data[(length & ^3)+2]) & 0xff) << 16
+	}
+	if extra >= 2 {
+		h ^= (uint32(data[(length & ^3)+1]) & 0xff) << 8
+	}
+	if extra >= 1 {
+		h ^= uint32(data[length & ^3]) & 0xff
+		h *= m
+	}
+
+	h ^= h >> 13
+	h *= m
+	h ^= h >> 15
+
+	return h
+}
+
+// crcpartitioner is a partitioner for crc32 we use it instead of just using sarama.NewCustomHashPartitioner(crc32.NewIEEE),
+// bo maintain compatibility, because the way that sarama distributes the result of the hash causes us to lose compatibility
+type crcpartitioner struct{}
+
+// Partition takes a message and partition count and chooses a partition
+// we made a custom partitioner, because
+func (crcpartitioner) Partition(msg *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	if numPartitions == 0 {
+		return 0, ErrNonPositivePartitions
+	}
+	// NOTE: the crc32 balancers in librdkafka don't differentiate between nil
+	//       and empty keys.  both cases are treated as unset.
```

Comment:
Same - does this match kapa's previous behaviour? I assume we weren't using librdkafka previously.

Reply:
The library we are using mimics librdkafka behavior and we promise librdkafka compat for it.